File: python/sim.py
import csv

import matplotlib.colors
from matplotlib.widgets import Button
from task import *
import task
import math
import numpy as np
import matplotlib
matplotlib.use("TkAgg")
from matplotlib.patches import FancyArrowPatch
import matplotlib.pyplot as plt
plt.ion()
from typing import Generator
import time

def pi_to_pi(x: float):
    return math.remainder(x, 2*math.pi)

class Sim:
    def __init__(self):
        self.cones = []
        self.visible_cones = []
        self.car_state = CarState(0, 0, 0)
        self.centerline = []

        self.fig, self.ax = plt.subplots()
        self.scatter = None
        self.arrow = FancyArrowPatch((0,0), (1,1), mutation_scale=20, color='blue', arrowstyle='->')
        self.plot_center_x = 0
        self.plot_center_y = 0
        self.ax.set_aspect('equal')
        plt.autoscale(False)

        self.zoom = 100
        self.do_auto_panning = True
        self.ax_button = plt.axes([0.65, 0.01, 0.3, 0.05])
        self.button = Button(self.ax_button, 'Toggle Auto Panning')
        self.ax_slider = plt.axes([0.1, 0.01, 0.45, 0.03])
        self.slider = plt.Slider(self.ax_slider, 'Zoom', 5, 200, valinit=100)
        self.slider.on_changed(self.update_zoom)
        self.button.on_clicked(self.toggle_auto_panning)
    
    def update_zoom(self, val):
        self.zoom = self.slider.val

    def toggle_auto_panning(self, event):
            self.do_auto_panning = not self.do_auto_panning

    def load_track_from_csv(self, file_path):
        with open(file_path, 'r') as file:
            reader = csv.reader(file, delimiter=",")
            self.cones = []
            for row in reader:
                if row[0] == "BLUE":
                    self.cones.append(Cone(Color.BLUE, float(row[1]), float(row[2])))
                elif row[0] == "YELLOW":
                    self.cones.append(Cone(Color.YELLOW, float(row[1]), float(row[2])))
                elif row[0] == "ORANGE":
                    self.cones.append(Cone(Color.ORANGE, float(row[1]), float(row[2])))
        
    def get_visible_cones(self) -> Generator[Cone, None, None]:
        for c in self.cones:
            distance_squared = (c.x-self.car_state.x)**2 + (c.y-self.car_state.y)**2
            if distance_squared  < 625 and distance_squared > 4:
                detection_angle = pi_to_pi(pi_to_pi(math.atan2(c.y-self.car_state.y, c.x-self.car_state.x)) - pi_to_pi(self.car_state.theta))
                if math.fabs(detection_angle) < 60/180*math.pi:
                    yield c

    def visualize(self):
        x_glob = [c.x for c in self.cones]
        y_glob = [c.y for c in self.cones]        

        x = [c.x for c in self.visible_cones]
        y = [c.y for c in self.visible_cones]
        color = [(0.0, 0.0, 1.0) if c.color==Color.BLUE else (1.0, 1.0, 0.0) if c.color==Color.YELLOW else (1.0, 0.5, 0.0) for c in self.visible_cones]

        self.ax.clear()
        self.scatter = self.ax.scatter(x, y, c=color, s=50)
        self.scatter = self.ax.scatter(x_glob, y_glob, s=10, c='gray')

        if self.car_state and self.do_auto_panning:
            car_x, car_y, car_theta = self.car_state.x, self.car_state.y, self.car_state.theta

            # Define the box size
            size = 15 / (self.zoom / 100)
            box_length = size * 1.8

            d_x = (box_length/2) * np.cos(car_theta)
            d_y = (box_length/2) * np.sin(car_theta)

            center_x = car_x + d_x
            center_y = car_y + d_y

            self.plot_center_x = 0.1 * center_x + 0.9 * self.plot_center_x
            self.plot_center_y = 0.1 * center_y + 0.9 * self.plot_center_y

            
            self.ax.set_xlim(self.plot_center_x - size, self.plot_center_x + size)
            self.ax.set_ylim(self.plot_center_y - size, self.plot_center_y + size)


        # Plot the car as an arrow
        if self.car_state:
            car_x, car_y, car_theta = self.car_state.x, self.car_state.y, self.car_state.theta
            car_length = 1.0  # Length of the car's arrow (visual representation)

            # Calculate the car's direction using trigonometry
            dx = car_length * np.cos(car_theta)
            dy = car_length * np.sin(car_theta)

            self.arrow.set_positions((car_x, car_y), (car_x + dx, car_y + dy))
            self.ax.add_patch(self.arrow)
            
        centerline_x = [c.x for c in self.centerline]
        centerline_y = [c.y for c in self.centerline]
        self.ax.plot(centerline_x, centerline_y, 'r-')


        plt.draw()
        plt.pause(0.01)

    def simulation_step(self):
        # Update car position
        if self.centerline is not None and len(self.centerline) >0:
            def find_point_on_centerline_with_distance(centerline, distance):
                total_distance = 0
                for i in range(1, len(centerline)):
                    p1 = centerline[i-1]
                    p2 = centerline[i]
                    segment_distance = math.hypot(p1.x - p2.x, p1.y - p2.y)
                    if total_distance + segment_distance > distance:
                        ratio = (distance - total_distance) / segment_distance
                        return Point2D(p1.x + (p2.x - p1.x) * ratio, p1.y + (p2.y - p1.y) * ratio)
                    total_distance += segment_distance
                return centerline[0]
            goal_point = find_point_on_centerline_with_distance(self.centerline, 1)
            direction_vector = [goal_point.x- self.car_state.x, goal_point.y - self.car_state.y]
            direction_norm = np.linalg.norm(direction_vector)
            
            goal_theta = math.atan2(direction_vector[1], direction_vector[0])
            max_theta_change = 0.5
            if pi_to_pi(goal_theta - self.car_state.theta) > max_theta_change:
                self.car_state.theta += max_theta_change
            elif pi_to_pi(goal_theta - self.car_state.theta) < -max_theta_change:
                self.car_state.theta -= max_theta_change
            else:
                self.car_state.theta = goal_theta
            self.car_state.theta = pi_to_pi(self.car_state.theta)
        self.car_state.x += np.cos(self.car_state.theta) * 0.5
        self.car_state.y += np.sin(self.car_state.theta) * 0.5


            

        self.visible_cones = list(self.get_visible_cones())
        # visible cones should be transformed here to the car's frame of reference
        transformed_cones = []

        for cone in self.visible_cones:
            # Translate cone position relative to the car
            dx = cone.x - self.car_state.x
            dy = cone.y - self.car_state.y

            # Rotate according to the car's heading (theta)
            rotated_x = dx * np.cos(-self.car_state.theta) - dy * np.sin(-self.car_state.theta)
            rotated_y = dx * np.sin(-self.car_state.theta) + dy * np.cos(-self.car_state.theta)
            transformed_cones.append(Cone(cone.color, rotated_x, rotated_y))

        self.centerline = task.calc_centerline(transformed_cones, self.car_state)
        


if __name__ == "__main__":
    sim = Sim()
    sim.load_track_from_csv("track.csv")
    
    while True:
        start_time = time.time()
        sim.simulation_step()
        sim.visualize()

        elapsed_time = time.time() - start_time
        time.sleep(max(0, 0.03 - elapsed_time))

File: cpp/sim.cpp
#include "matplotlibcpp.h"
#include <cmath>
#include <fstream>
#include <iostream>
#include <memory>
#include <sstream>
#include <vector>

#include "task.h"

namespace plt = matplotlibcpp;

class Sim {
public:
  Sim() : car_state(0, 0, 0) { 
    plt::ion(); 
    // use tkagg
    plt::backend("TkAgg");
    car_state = CarState(0, 0, 0);
  }

  void loadTrackFromCSV(const std::string &file_path) {
    std::ifstream file(file_path);
    std::string line;
    cones.clear();

    if (!file.is_open()) {
      std::cerr << "Failed to open file: " << file_path << std::endl;
      return;
    }

    while (std::getline(file, line)) {
      std::istringstream stream(line);
      std::string color, x_str, y_str;

      // Read color, x, and y as comma-separated values
      if (!std::getline(stream, color, ',') ||
          !std::getline(stream, x_str, ',') ||
          !std::getline(stream, y_str, ',')) {
        std::cerr << "Error parsing line: " << line << std::endl;
        continue; // Skip this line if parsing fails
      }

      // Convert x and y from strings to doubles
      double x, y;
      try {
        x = std::stod(x_str);
        y = std::stod(y_str);
      } catch (const std::invalid_argument &) {
        std::cerr << "Invalid coordinates in line: " << line << std::endl;
        continue;
      }

      // Determine the color and add to cones
      if (color == "BLUE") {
        cones.emplace_back(Color::BLUE, x, y);
      } else if (color == "YELLOW") {
        cones.emplace_back(Color::YELLOW, x, y);
      } else if (color == "ORANGE") {
        cones.emplace_back(Color::ORANGE, x, y);
      } else {
        std::cerr << "Unknown color: " << color << " in line: " << line
                  << std::endl;
      }
    }

    file.close();
    std::cout << "Loaded " << cones.size() << " cones from CSV."
              << std::endl; // Final confirmation
  }

  void simulationStep() {
    if (centerline.size() >= 1) {
      // Update car's position based on the centerline points
      float direction_x = centerline[0].x - car_state.x;
      float direction_y = centerline[0].y - car_state.y;
      float distance =
          std::sqrt(direction_x * direction_x + direction_y * direction_y);

      car_state.x += 0.1 * direction_x / distance;
      car_state.y += 0.1 * direction_y / distance;
      car_state.theta = std::atan2(direction_y, direction_x);
    }

    visible_cones = getVisibleCones();
    std::vector<Cone> transformed_cones = transformConesToCarFrame();

    // Compute the centerline (this function needs to be implemented)
    centerline = calcCenterline(transformed_cones, car_state);
  }

  void visualize() {
    // Clear the previous plot
    plt::clf();

    // Plot visible cones
    for (const auto &cone : visible_cones) {
      std::string color;
      if (cone.color == Color::BLUE)
        color = "blue";
      else if (cone.color == Color::YELLOW)
        color = "yellow";
      else
        color = "orange";

      plt::plot({cone.x}, {cone.y}, {{"marker", "o"}, {"color", color}});
    }

    // Plot the car's position as an arrow
    double car_length = 1.0;
    double dx = car_length * std::cos(car_state.theta);
    double dy = car_length * std::sin(car_state.theta);
    // plt::arrow(car_state.x, car_state.y, dx, dy, 0.5, {{"color", "blue"}});

    // Plot the centerline
    if (!centerline.empty()) {
      std::vector<double> cx, cy;
      for (const auto &point : centerline) {
        cx.push_back(point.x);
        cy.push_back(point.y);
      }
      plt::plot(cx, cy, "r-");
    }
    std::cout << "Car position: (" << car_state.x << ", " << car_state.y
              << "), theta: " << car_state.theta << std::endl;

    unsigned box_length = 25;

    float d_x = (box_length/2) * std::cos(car_state.theta);
    float d_y = (box_length/2) * std::sin(car_state.theta);

    float center_x = car_state.x + d_x;
    float center_y = car_state.y + d_y;    

    plt::xlim(center_x - 15, center_x + 15);
    plt::ylim(center_y - 15, center_y + 15);

    // Draw and pause to create animation effect
    plt::pause(0.1);
  }

private:
  std::vector<Cone> cones;
  std::vector<Cone> visible_cones;
  std::vector<Point2D> centerline;
  CarState car_state;

  double piToPi(double angle) { return std::remainder(angle, 2 * M_PI); }

  std::vector<Cone> transformConesToCarFrame() {
    std::vector<Cone> transformed;
    for (const auto &cone : visible_cones) {
      // Translate cone position relative to the car
      double dx = cone.x - car_state.x;
      double dy = cone.y - car_state.y;

      // Rotate based on car's orientation (theta)
      double rotated_x =
          dx * std::cos(-car_state.theta) - dy * std::sin(-car_state.theta);
      double rotated_y =
          dx * std::sin(-car_state.theta) + dy * std::cos(-car_state.theta);

      transformed.emplace_back(cone.color, rotated_x, rotated_y);
    }
    return transformed;
  }

  std::vector<Cone> getVisibleCones() {
    std::vector<Cone> visible_cones;
    for (const auto &cone : cones) {
      double dist_squared =
          std::pow(cone.x - car_state.x, 2) + std::pow(cone.y - car_state.y, 2);
      if (dist_squared < 625) { // Within visible range
        double detection_angle = piToPi(
            piToPi(std::atan2(cone.y - car_state.y, cone.x - car_state.x)) -
            piToPi(car_state.theta));
        if (std::fabs(detection_angle) < M_PI / 3) { // Within 60 degrees
          visible_cones.push_back(cone);
        }
      }
    }
    return visible_cones;
  }
};

int main() {
  Sim sim;
  sim.loadTrackFromCSV("track.csv");

  // Main loop for the simulation
  while (true) {
    sim.simulationStep();
    sim.visualize();
  }

  return 0;
}
